models/issue_watch.go

```diff
@@ -5,86 +5,133 @@
+	IssueWatchModeNone IssueWatchMode = iota // 0
+	// IssueWatchModeNormal watch issue (from other sources)
+	IssueWatchModeNormal // 1
+	// IssueWatchModeDont explicit don't auto-watch
+	IssueWatchModeDont // 2
+	// IssueWatchModeAuto watch issue (from AutoWatchOnChanges)
+	IssueWatchModeAuto // 3
+)
+
 // IssueWatch is connection request for receiving issue notification.
 type IssueWatch struct {
 	ID          int64              `xorm:"pk autoincr"`
 	UserID      int64              `xorm:"UNIQUE(watch) NOT NULL"`
 	IssueID     int64              `xorm:"UNIQUE(watch) NOT NULL"`
-	IsWatching  bool               `xorm:"NOT NULL"`
+	Mode        IssueWatchMode     `xorm:"NOT NULL DEFAULT 1"`
 	CreatedUnix timeutil.TimeStamp `xorm:"created NOT NULL"`
 	UpdatedUnix timeutil.TimeStamp `xorm:"updated NOT NULL"`
 }
 
 // IssueWatchList contains IssueWatch
 type IssueWatchList []*IssueWatch
 
-// CreateOrUpdateIssueWatch set watching for a user and issue
-func CreateOrUpdateIssueWatch(userID, issueID int64, isWatching bool) error {
-	iw, exists, err := getIssueWatch(x, userID, issueID)
+// CreateOrUpdateIssueWatchMode set IssueWatchMode for a user and issue
+func CreateOrUpdateIssueWatchMode(userID, issueID int64, mode IssueWatchMode) error {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+	err := createOrUpdateIssueWatchMode(sess, userID, issueID, mode)
+	if err != nil {
+		return err
+	}
+	return sess.Commit()
+}
+
+func createOrUpdateIssueWatchMode(e Engine, userID, issueID int64, mode IssueWatchMode) error {
+	iw, exists, err := getIssueWatch(e, userID, issueID)
 	if err != nil {
 		return err
 	}
 
-	if !exists {
+	if !exists && mode != IssueWatchModeNone {
 		iw = &IssueWatch{
-			UserID:     userID,
-			IssueID:    issueID,
-			IsWatching: isWatching,
+			UserID:  userID,
+			IssueID: issueID,
+			Mode:    mode,
 		}
-
-		if _, err := x.Insert(iw); err != nil {
+		if _, err = e.Insert(iw); err != nil {
 			return err
 		}
 	} else {
-		iw.IsWatching = isWatching
-
-		if _, err := x.ID(iw.ID).Cols("is_watching", "updated_unix").Update(iw); err != nil {
+		if mode != IssueWatchModeNone {
+			iw.Mode = mode
+			if _, err = e.ID(iw.ID).Cols("updated_unix", "mode").Update(iw); err != nil {
+				return err
+			}
+		}
+		if _, err = e.ID(iw.ID).Delete(iw); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
 // GetIssueWatch returns all IssueWatch objects from db by user and issue
-// the current Web-UI need iw object for watchers AND explicit non-watchers
 func GetIssueWatch(userID, issueID int64) (iw *IssueWatch, exists bool, err error) {
 	return getIssueWatch(x, userID, issueID)
 }
 
-// Return watcher AND explicit non-watcher if entry in db exist
 func getIssueWatch(e Engine, userID, issueID int64) (iw *IssueWatch, exists bool, err error) {
 	iw = new(IssueWatch)
 	exists, err = e.
 		Where("user_id = ?", userID).
 		And("issue_id = ?", issueID).
+		And("mode <> ?", IssueWatchModeNone).
```

Comment:
Do you need to exclude `IssueWatchModeDont`?

Reply:
If not add a comment to the comments to explain that the results explicitly include those who have elected not to watch an issue.

Reply:
I used enum 0 = IssueWatchNone to represent entrys witch didnt migrate corektly and as delete type.

I can remove this type and drop non correct entrys on migration & create a own deletion function for issuewatch db-entrys?